File: LeetCode/Validate Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
 public:
  bool check(TreeNode* node, int* small, int* big) {
    if (!node) return true;

    if (small != NULL && node->val <= *small) return false;
    if (big != NULL && node->val >= *big) return false;

    return check(node->left, small, &node->val) &&
           check(node->right, &node->val, big);
  }

  bool isValidBST(TreeNode* root) { return check(root, NULL, NULL); }
};

File: LeetCode/Validate Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root: 
            return True
        
        stack = []
        node = root
        prev = None
        
        while (node != None or len(stack) > 0):
            while(node != None):
                stack.append(node)
                node = node.left
                
            node = stack.pop()
                
            if (prev != None and prev.val >= node.val):
                return False
            
            prev = node
            node = node.right
        
        return True
            